src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -12,18 +12,30 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering::SeqCst;
+
 use risingwave_hummock_sdk::key::{Epoch, FullKey};
 use risingwave_pb::hummock::SstableInfo;
 use tokio::task::JoinHandle;
 
+use super::{
+    InMemSstableWriter, InMemWriterBuilder, SstableMeta, SstableWriter, SstableWriterBuilder,
+    StreamingSstableWriter, StreamingWriterBuilder,
+};
 use crate::hummock::sstable_store::SstableStoreRef;
 use crate::hummock::utils::MemoryTracker;
 use crate::hummock::value::HummockValue;
-use crate::hummock::{CachePolicy, HummockResult, SstableBuilder};
+use crate::hummock::{
+    CachePolicy, HummockResult, MemoryLimiter, SstableBuilder, SstableBuilderOptions,
+};
 
 #[async_trait::async_trait]
-pub trait TableBuilderFactory {
-    async fn open_builder(&self) -> HummockResult<(MemoryTracker, SstableBuilder)>;
+pub trait TableBuilderFactory<B>: Send
```

Comment:
We may use `B` as the associate type of `TableBuilderFactory`, and then we don't need to specify a new generic type parameter of `SstableWriterBuilder` whenever we use `TableBuilderFactory`.

Reply:
`TableBuilderFactory` -> `SstableWrtierBuilder` is a one-to-many relationship. For example, `RemoteTableBuilderFactory` may have `InMemSstableWriter` or `StreamingUploadSstableWriter` as `B`.

Reply:
Using `B` as an associate type has no conflict with the one-to-many relation.

What I meant was to make the trait into
```
pub trait TableBuilderFactory {
    type WriterBuilder: SstableWriterBuilder;
    async fn open_builder(&self) -> HummockResult<(MemoryTracker, SstableBuilder<WriterBuilder::Writer>)>;
}
```

The current definition of `CapacitySplitSstableBuilder` is
```
pub struct CapacitySplitTableBuilder<F, B, S>
where
    F: TableBuilderFactory<B>,
    B: SstableWriterBuilder,
    S: SstableBuilderSealer<B::Writer>
```
After we use associated type, it can be
```
pub struct CapacitySplitTableBuilder<F, S>
where
    F: TableBuilderFactory,
    S: SstableBuilderSealer<F::WriterBuilder::Writer>,
```

Reply:
Then how do I use `RemoteTableBuilderFactory<InMemWriterBuilder>` and `RemoteTableBuilderFactory<StreamingUploadWriterBuilder>` at the same time?

Reply:
`RemoteTableBuilderFactory` can still take a generic parameter `<B: SstableWriterBuilder>`.

```
impl<B: SstableWriterBuilder> TableBuilderFactory for RemoteTableBuilderFactory<B> {
    type WriterBuilder = B;
    ...
}
```

Reply:
I see. It's me being lame with rust 🤣